### DISCUSSION TITLE
How Do I Return 444 And Drop Connections for Clients of Hosts not in ACL CIDR Range? If it’s not possible, I submit it as a feature request.

### DISCUSSION THREAD

**Discussion by charlestephen** [ZoeyVid/NPMplus#1102] [Q&A] [answered]
I have a server that currently functions as my Plex Media Server which I share with others and thus has some ports open to the world. I also secure all of my internal services with TLS termination using a combination of Acme.sh and NPMPlus. Despite the advanced configs with CrowdSec and ModSec, I’m still getting hit with thousands of attempts daily trying to either scan my network (which I’m ok with) or pull credentials/do SQL injection and I’ve noticed it slows down my main server significantly.

I cut down by half the attempts by securing a ZeroSSL certificate for my IP address and importing it to NPMPlus as a redirection host. Now anyone calling my IP address is automagically offloaded to Google. While that has significantly improved my server, I’d load, I’d like to do more both for energy and security sake.

Basically, I don’t want hosts on Access Lists to return 401, 403, or 404 error data to clients not matching the CIDR ranges. I just want to issue a 444 and drop the connection. This has the benefit of not informing the client why or if it’s requests would have returned legitimate results, thus cutting down on people constantly pinging my firewall.

One way I’ve figured out how to get this work is putting the following in the the http_top.conf file:

`location /  {
        allow 192.168.0.0/16;
        deny all;
        return 444;
    }`

But this seems to apply the rule to all hosts. Is there a way to be more selecting?  If I do custom location (not an advanced configuration), with that same information, should it work or will I simply break NPMPlus? I haven’t had luck and have tried this on multiple hosts. Am I missing some combination of a custom default page and a 404 redirection host or something that achieves this? Or is the only way to do it to add individual configs and I’m doing it wrong?

If it’s currently not possible, I’d like to suggest it as a feature request: return 444 and drop connection for clients not in the CIDR ranges of hosts with access lists.

**Comment by Zoey2936** [comment]
`location /  { allow 192.168.0.0/16; deny all; return 444; }`
does this not basically mean: return 403 for all connection except it is from 192.168.0.0/16 then instead drop the connection?

**Comment by Zoey2936** [comment]
also it should crash:
http_top.conf mean it is inside the http directive, but location can not be in http, only server or location: https://nginx.org/en/docs/http/ngx_http_core_module.html#location

**Comment by Zoey2936** [comment, marked as answer]
sorry, but I think it is not possible with (free)nginx to change the response code from 403 to something else. if you want you can write an email to the freenginx mailing list and ask if they can add this as an option

**Comment by Zoey2936** [comment]
I would recommend you to this instead on firewall level, like using ufw, and only allowing ips you want

**Reply by charlestephen** [reply]
Well, the problem is I have a hosted server that is open to the public which is secured (it’s Plex) I just don’t want the other hosts servered by this instance of NPMPlus or other instances I have internally to cause my default install to return any details and I can’t control from where people connect to the allowed server, thus why I was trying to minimize my exposure through perfidy.

**Reply by Zoey2936** [reply]
you can set ip bindings via env in npmplus, so ports are not exposed to public

**Reply by charlestephen** [reply]
I don’t think you’re quite underistanding me fully: this NPMPlus instance is on a server who primarily faces the public but I’m trying to minimize how much information can be gleaned about my internal infrastructure by returning the same command for all requests that aren’t from internal cidr ranges.

Could I possibly use the `geo` directive to deny external IPs?

Something maybe like this: 

```
geo $deny_external {
  default 1;
  # Allow internal CIDRs
  192.168.2.0/24 0;
  10.89.0.0/24 0;
  100.100.101.0/24 0;
}

server {
  if ($deny_external) {
    return 444;
  }
  set $forward_scheme https;
  set $server "192.168.2.2";
  set $port 9090;

  listen unix:/run/nginx-6.sock;

  listen 0.0.0.0:80;
  listen [::]:80;


  listen 0.0.0.0:443 ssl;
  listen [::]:443 ssl;

  listen 0.0.0.0:443 quic;
  listen [::]:443 quic;


  more_set_headers 'Alt-Svc: h3=":443"; ma=86400';


  server_name internal.domain.tldf;


  # Certbot TLS
  include conf.d/include/tls-ciphers.conf;
  ssl_certificate /data/tls/certbot/live/npm-6/fullchain.pem;
  ssl_certificate_key /data/tls/certbot/live/npm-6/privkey.pem;
  ssl_trusted_certificate /data/tls/certbot/live/npm-6/chain.pem;


  more_clear_headers "Expect-CT";
  include conf.d/include/hsts.conf;


  # Force TLS
  include conf.d/include/force-tls.conf;


  # Enable Brotli
  include conf.d/include/brotli.conf;


  # Access Rules: 4 total
  allow 192.168.2.0/24;

  allow 192.168.5.0/24;

  allow 100.100.101.0/24;

  allow 10.89.0.0/24;

  deny all;

  # Access checks must...

  satisfy any;


  modsecurity on;
  modsecurity_rules_file /usr/local/nginx/conf/conf.d/include/modsecurity-crs.conf;


  include conf.d/include/always.conf;


  # custom locations
  location / {
    include conf.d/include/always.conf;


    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;


    # Proxy!
    include conf.d/include/proxy.conf;
  }


  # Custom
  include /data/nginx/custom/server_proxy.conf;
  }# Other server block settings
}
```

**Reply by Zoey2936** [reply]
you want to block all request which are not from your private ip ranges, right?

**Reply by charlestephen** [reply]
I’d like for hosts configured as private only to have requests refused/dropped when done from public clientss. Conversely, I want for both public and private clients to have their connections fulfilled when reaching public hosts.

**Reply by Zoey2936** [reply]
you can setup two npmplus instances, one with ip bindings to your public ip and a second npmplus bound to a private ip

**Reply by Zoey2936** [reply]
network level protection is in any case better then nginx configuration level

**Reply by charlestephen** [reply]
Youre probably right.

**Reply by Zoey2936** [reply]
I have the same setup for me. I have two NPMplus one serving public and one private and if I want hosts on both, I just configure them on both

**Comment by Zoey2936** [comment]
a different question, since I want to further improve NPMplus in ssl/tls questions: why do you use acme.sh and zerossl instead of built in certbot (where you can also configure it to use zerossl). And is there a reason for zerossl, only becaus eit is the default of acme.sh?

**Reply by charlestephen** [reply]
I use Acme.sh for things that can't support containers or where it would be too difficult/timely to install and maintain like my printer (the former) and my router, which has RADIUS and VPN servers on it (the latter). Everytime there’s a firmware update on my router, it wipes out most installd applications but Acme.sh is easy to redownload, run, and has a hook for my router (UniFi).

I used ZeroSSL through their web portal cause they were the only free CA that I could find willing to issue a cert for an IP address and you can’t do it via automated methods. I needed that cert to seamlessly pass incoming HTTP and HTTPS traffic on my WAN IP to other endpoints like Google.com, thus redirecting a lot of potentially malicious traffic and reducing my attack vectors.

Since I implemented that change, I’ve noticed my server using dramatically less memory and my connection is much faster/lower latency than it had been. I’m definitely going to renew the IP cert when it comes time and continue passing IP-only traffic to public endpoints.

**Reply by Zoey2936** [reply]
Thanks! So it more a question of preference for acme.sh (it does not require python - which why I want to change NPMplus to it in the future). And ZeroSSL they provicde certs for ips? I didn't know that. Do you if they are also doing that via acme? I for myself currently use google public CA, but I also want to move back to Letsencrypt or ZeroSSL, since google has no ocsp support but letsencrypt want to stop proving it because it is tooo expensive for them. So I think NPMplus will also provice ZeroSSL as default in the future

**Reply by charlestephen** [reply]
I prefer Acme.sh myself as well. And I’m welcome you changing NPMPlus to it in future. Yes, ZeroSSL provides IP certs for free but not via acme. I do believe a paid account may get you acme compatibility but I’m unsure. The website said web authentication with a .well-known file was the only method of confirming my IP. I didn’t know LetsEncrypt doesn’t want to provide OCSP support. That stinks,  my entire domain is on them and I have CAA DNS records locking it so. Rats.

**Reply by Zoey2936** [reply]
Yes, see here:
- https://letsencrypt.org/2024/07/23/replacing-ocsp-with-crls
- https://github.com/caddyserver/caddy/issues/6282
- https://community.letsencrypt.org/t/sunsetting-of-ocsp-in-favor-of-older-technology/222589/19
- https://community.letsencrypt.org/t/what-will-happen-to-must-staple/222397

It is very sad to see, since CRLS would be too big to cover all certs and are never up-to-date, so a very bad replacement. I can understand that they are dropping it because of the costs they have with it, but yes... CRLS is not a replacement at all.
